### ISSUE TITLE
Multiple Authorize Attributes in MinAPI generates "AuthenticationScheme: Bearer was forbidden."

### ISSUE THREAD

**Issue by johnkwaters** [dotnet/aspnetcore#47581]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

I have previously successfully added a custom AuthorizatoinAttribute to a MinAPI endpoint.
Basically, I took the attribute and added it using

```
        builder.Add(endpointBuilder => { endpointBuilder.Metadata.Add(
            new FeatureAuthorizeAttribute(features)); });
```

Where builder is the Min API IEndPointConventionBuilder, so this just adds this attribute (my FeatureAuthorizeAttribute).

From there, it works (I'll spare you the details but it involves an associated Requirement class, RequirementHandler and CustomAuthorizationPolicyProvider)

However I now need to add a second custom Authorize attribute. (a ScopeAuthorizeAttribute). 

```
        builder.Add(endpointBuilder => {
            endpointBuilder.Metadata.Add(
                new ScopeAuthorizeAttribute(Scopes));
        });
```

This works in my ApiController based code - you then need both these - the correct Scope claim and correct Fetaures). But in Min API, this doesn't not work. The requirement handler is never even called, instead, something generates a 403 with

[19:15:26 INF] Authorization failed. These requirements were not met:
TrailheadTechnology.API.Auth.ScopeRequirement
[19:15:26 INF] AuthenticationScheme: Bearer was forbidden.
[19:15:26 INF] Setting HTTP status code 403.

There is only one AuthenticationSchems, Bearer - I don't know why Min API thinks the token is not valid when the endpoint has two different auth attributes - neither of which specify a scheme.

Thoughts?

### Expected Behavior

The code should recognize the caller is authenticated and pass through to the correct policy handler

### .NET Version

7

**Comment by Tratcher** [comment]
The custom authorize attribute approach is discouraged. We recommend using authorization policies instead.

We'd need to see your custom attribute implementations to investigate this.

**Comment by johnkwaters** [comment]
Well under the hood this is actually a dynamically generated authorization policy, Below ii how the pieces fit together, and I'd love to see how you could do that in MinAPI without custom auth policies, but why is it complaining about Bearer not being allowed as soon as I attach a second attribute, but is fine with 1, and ApiControllers are fine with 2. Note the whole purpose of this somewhat convoluted construct is to be able to decorate an endpoint with which features must be present in the usercontext of the caller for the endpoint to be called. Happy for cleaner paths. this has served us well for many years, but is clunky. Note we need several such policies in parallel - i.e. all endpoints in a group may need ScopeRequirement (scope policy) x, whereas different methods also feature policy y or z or y,z.

1. The auth attribute (stores its ctor param, an array of Feature enums, as a dynamic policyname containing the features, for instance "Feature_x,y,z".

```
internal class FeatureAuthorizeAttribute : AuthorizeAttribute
{
    const string FEATURE_POLICY_PREFIX = "Feature";

    public FeatureAuthorizeAttribute(params Feature[] features) => Features = features;

    public Feature[] Features
    {
        get
        {
            var policyFeature = Policy;
            // ReSharper disable once CollectionNeverQueried.Local 
            // it is queried but not in the this context
            var features = new List<Feature>();
            foreach (var fp in policyFeature!.Substring(FEATURE_POLICY_PREFIX.Length + 1).Split(','))
            {
                var tryParse = Enum.TryParse<Feature>(fp, out var feature);
                if (tryParse)
                {
                    features.Add(feature);
                }
            }

            return Array.Empty<Feature>();
        }

        set => Policy = $"{FEATURE_POLICY_PREFIX}_{string.Join(',', value)}";
    }
}

```

2. The custom policy provider, that when passed a string like "Feature_x,y,z" turns it into a policy with a FeatureRequirement, passing the features to it's ctor (i.e. bouncing back from serialized string of features to the actual features!)

```
public class CustomPolicyProvider : IAuthorizationPolicyProvider
{
    const string FEATURE_POLICY_PREFIX = "Feature";

    /// <summary>
    /// FallBack policy 
    /// </summary>
    private DefaultAuthorizationPolicyProvider FallbackPolicyProvider { get; }

    /// <summary>
    /// Constructor
    /// </summary>
    /// <param name="options">options</param>
    public CustomPolicyProvider(IOptions<AuthorizationOptions> options)
    {
        // ASP.NET Core only uses one authorization policy provider, so if the custom implementation
        // doesn't handle all policies (including default policies, etc.) it should fall back to an
        // alternate provider.
        FallbackPolicyProvider = new DefaultAuthorizationPolicyProvider(options);
    }

    /// <summary>
    /// Fallback policy
    /// </summary>
    /// <returns>The fallback policy of the FallbackPolicyProvide</returns>
    public Task<AuthorizationPolicy> GetFallbackPolicyAsync() =>
        FallbackPolicyProvider.GetFallbackPolicyAsync()!;

    /// <summary>
    /// Default policy
    /// </summary>
    /// <returns>The default policy of the FallbackPolicyProvide</returns>
    public Task<AuthorizationPolicy> GetDefaultPolicyAsync() =>
        FallbackPolicyProvider.GetDefaultPolicyAsync();

    /// <summary>
    /// Searches for a policy with the given name
    /// Policies are looked up by string name, so expect 'parameters' (like age)
    /// to be embedded in the policy names. This is abstracted away from developers
    /// by the more strongly-typed attributes derived from AuthorizeAttribute
    /// (like [MinimumAgeAuthorize()] in this sample)
    /// https://docs.microsoft.com/en-us/aspnet/core/security/authorization/iauthorizationpolicyprovider
    /// </summary>
    /// <param name="policyName">The name to search for</param>
    /// <returns>A  matching policy, if found</returns>
    public Task<AuthorizationPolicy> GetPolicyAsync(string policyName)
    {
        var isFP = policyName.StartsWith(FEATURE_POLICY_PREFIX, StringComparison.OrdinalIgnoreCase);

        if (isFP)
        {
            var features = new List<Feature>();
            foreach (var fp in policyName.Substring(FEATURE_POLICY_PREFIX.Length + 1).Split(','))
            {
                var tryParse = Enum.TryParse<Feature>(fp, out var feature);
                if (tryParse)
                {
                    features.Add(feature);
                }
            }

            var policy = new AuthorizationPolicyBuilder();
            policy.AddRequirements(new FeatureRequirement(features.ToArray()));
            return Task.FromResult(policy.Build());
        }

        return Task.FromResult<AuthorizationPolicy>(null!);
    }
}
```

3. The Feature Requirement that holds on to those features:

```
public class FeatureRequirement : IAuthorizationRequirement
{
    /// <summary>
    /// Required feature
    /// </summary>
    public Feature[] RequiredFeatures { get; set; }

    /// <summary>
    /// Constructor
    /// </summary>
    /// <param name="requiredFeatures"></param>
    public FeatureRequirement(Feature[] requiredFeatures)
    {
        RequiredFeatures = requiredFeatures;
    }
}

```

5. The FeatureHandler that compares the features in the calling usercontext to those in the requirement

```
/// <summary>
/// Authorization handler that checks access to a given feature
/// </summary>
public class FeatureHandler : AuthorizationHandler<FeatureRequirement>
{
    private readonly IUserContext _userContext;

    /// <summary>
    /// Constructor
    /// </summary>
    /// <param name="userContext">User context</param>
    public FeatureHandler(IUserContext userContext)
    {
        _userContext = userContext;
    }

    /// <summary>
    /// Check if the user has the feature required by the requirement
    /// </summary>
    /// <param name="context">Authorization context, call Succeed if OK</param>
    /// <param name="requirement">Contains the feature requirement</param>
    /// <returns></returns>
    protected override async Task HandleRequirementAsync(AuthorizationHandlerContext context,
        FeatureRequirement requirement)
    {
        if (context.User.Identity?.IsAuthenticated ?? false)
        {
            if (_userContext.Features.Any(t => requirement.RequiredFeatures.Contains(t)))
            {
                context.Succeed(requirement);
            }            
        }
    }
}


```

**Comment by Tratcher** [comment]
Triage: I won't be able to look into this further, un-assigning.

**Comment by johnkwaters** [comment]
Anyone else want to chime in here? Basically adding two Authorize-based attributes doesn't work... and I am interested to use a more modern policy based approach - still with multiple policies - but need pointers.

**Comment by johnkwaters** [comment]
@Tratcher I rewrote this code to NOT use AuthorizeAtttributes but I get the exact same error. If I add TWO different policies to a MinAPI endpoint, even though both are fulfilled, what happens is the first one gets evaluated and returns success, then the second policy name is resolved to the correct requirement, but instead of (before) calling the handler, it short circuits out with 

[14:08:44 INF] Authorization failed. These requirements were not met:
TrailheadTechnology.API.Auth.ScopeRequirement
[14:08:44 INF] AuthenticationScheme: Bearer was forbidden.

In other words it seems that the second policy is not called due to the framework deciding that the Bearer token, which was fine for the first policy, is no longer usable for the second one.

Can you please verify that you can have more than one policy for an end point? I can't practically create composite policies... you could have a Controller and a Method level Authorize attribute in AspNetCore controllers...

**Comment by johnkwaters** [comment]
Solved!

My code wasn't registering the ScopeHandler for that custom ScopeRequirement! Red herring!

```
        srv.AddScoped<IAuthorizationHandler, FeatureHandler>();
        srv.AddScoped<IAuthorizationHandler, ScopeHandler>();
        srv.AddSingleton<IAuthorizationPolicyProvider, CustomPolicyProvider>();
```
